### PR TITLE
feat: enabled overriding any day

### LINKED ISSUE #6279: Date overrides don't set you as available if the regular weekday has start and end time disabled
Found a bug? Please fill out the sections below. 👍

### Issue Summary

If you turn off your availability for an entire day, the Date overrides do not show you as available on that day.

### Steps to Reproduce

1. Go to `Availability` in the sidebar
2. Select calendar
3. Turn off all the days, so that the start and end time boxes no longer show
4. Add date overrides for days that are turned off
5. Visit public page, and tried to book a meeting

### Actual Result
The date overrides do not show up on the calendar, so it tells you "No availability in January"

### Expected Result
The date overrides should override your availability even if the regular day availability is turned off.

### Why is this considered a bug

I consider this a bug, because the Date overrides do not actually override your availability (and in this case, your lack of availability) for the days that are turned off.

This makes it impossible to allow booking of one-time weekend meetings without enabling regular weekend availability (usually not desirable). (Same thing for other non-weekend days, which you are usually not available on, but want to allow a one-time booking on that weekday).

### Technical details
- Firefox 108.0.1
- Using https://app.cal.com

### LINKED ISSUE #6311: [CAL-697] allow date overrides when all availability is off
### Is your proposal related to a problem?

Yes. Cal.com is almost an event booking system. One tweak and it could be. For example, if I want to host 3 different event types, each representing a different cuisine cooking class, I can make an availability matrix for each. Likely, I would turn off all availability, and add date overrides for each scheduled event. This would allow people to book for only specific days.

Currently, if all availability is switched off, turning on an override does nothing last time I checked. Correct me if I'm wrong.

As a user's business grows, it would likely turn into a more permanent recurring calendar, thus allow the user to continue using Cal.com for its intended purpose of booking up a calendar of slots.

One other thing to note, is that it would be nice for each event type to have its own conflict calendar for this same purpose. Otherwise, there's no way to host discrete classes on specific datetimes and not have them battle eachother with multiple global conflict calendars.

### Describe alternatives you've considered

In order to meet this need, we turned to TicketTailor, but ultimately Cal.com could be a good fit for this use case.

<sub>[CAL-697](https://linear.app/calcom/issue/CAL-697/allow-date-overrides-when-all-availability-is-off)</sub>

### PR DESCRIPTION
**Pull request by emrysal**
## What does this PR do?

Stop caring about the workingHours to create a date override.

Resolves #6311
Resolves #6279

image

image

### CODE REVIEW

**Comment by PeerRich** [comment]
wait does this enable adding days? or is that another PR

File: packages/trpc/server/routers/viewer/availability.tsx [hunk 1]
```diff
@@ -54,76 +54,6 @@ export const availabilityRouter = router({
     .query(({ input }) => {
       return getUserAvailability(input);
     }),
-  defaultValues: authedProcedure.input(z.object({ scheduleId: z.number() })).query(async ({ ctx, input }) => {
```
Comment (zomars) [inline packages/trpc/server/routers/viewer/availability.tsx:57]: Is this not needed anymore?

Comment (emrysal) [reply packages/trpc/server/routers/viewer/availability.tsx:57]: I'm not sure why I built this endpoint and didn't just extend get: - I revisited the code and I couldn't tell why I did this so I removed it.

**Review by zomars — CHANGES_REQUESTED** [review]
Blocking to solve the day off override mystery.


image

image

File: packages/core/getUserAvailability.ts [hunk 2]
```diff
@@ -256,8 +256,8 @@ export async function getUserAvailability(
       const startTime = dayjs.utc(override.startTime);
       const endTime = dayjs.utc(override.endTime);
       return {
-        start: dayjs.utc(override.date).hour(startTime.hour()).minute(startTime.minute()).toDate(),
-        end: dayjs.utc(override.date).hour(endTime.hour()).minute(endTime.minute()).toDate(),
+        start: dayjs.tz(override.date, timeZone).hour(startTime.hour()).minute(startTime.minute()),
```
Comment (CarinaWolli) [inline packages/core/getUserAvailability.ts:259]: My date overrides are shown on a wrong day. I have organizer and booker in the same timezone (Eastern Time). The date override is always applied one day before the actual day. 

It looks like setting the hour and minute changes the date to the day before

**Review by CarinaWolli — CHANGES_REQUESTED** [review]
My date overrides are shown shown on the wrong day (more details to it in the comment below): 

Screenshot 2023-01-19 at 14 23 58

That's where it is applied, but should be applied for 27 January
Screenshot 2023-01-19 at 14 24 03

Screenshot 2023-01-19 at 14 24 08

**Comment by mrkirchner** [comment]
Whats the progress on this its been open for a while because it may fix a few daylight saving issues.

**Comment by emrysal** [comment]
@mrkirchner Big push pending finalisation that handles the busyTimes & also the round robin fixed/collective attendees.

**Comment by zomars** [comment]
Setting these overrides 

image

Works for available to unavailable days but I cannot override unavailable to available days

image

File: apps/web/components/booking/pages/AvailabilityPage.tsx [hunk 3]
```diff
@@ -61,30 +63,45 @@ const useSlots = ({
...
+  }>({});
+  const utils = trpc.useContext();
   const { data, isLoading, isPaused } = trpc.viewer.public.slots.getSchedule.useQuery(
     {
       eventTypeId,
       eventTypeSlug,
       usernameList,
       startTime: startTime?.toISOString() || "",
       endTime: endTime?.toISOString() || "",
-      timeZone,
```
Comment (zomars) [inline apps/web/components/booking/pages/AvailabilityPage.tsx:71]: Do we not need timeZone anymore?

File: apps/web/components/booking/pages/AvailabilityPage.tsx [hunk 4]
```diff
@@ -61,30 +63,45 @@ const useSlots = ({
...
+      });
+      setGroupedByDate({
+        ...groupedByDate,
+        ...newGroupedByDate,
+      });
     }
-  }, [data]);
-
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [data?.slots]);
```
Comment (zomars) [inline apps/web/components/booking/pages/AvailabilityPage.tsx:102]: Can we compute this on the server?

File: packages/lib/server/getAvailability.ts [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+  dateTo,
+}: {
+  timeZone?: string;
+  availability: { startTime: Date; endTime: Date; date: Date | null; days: number[] }[];
+  dateTo: Date;
+  dateFrom: Date;
+}) => {
+  const workingDates = availability
+    .filter((value) => typeof value.days !== "undefined")
+    .reduce((dates, block) => {
```
Comment (zomars) [inline packages/lib/server/getAvailability.ts:29]:
```suggestion
    .reduce((dates, block) => {
      if (typeof block.days === "undefined") return dates;
```

Comment (zomars) [reply packages/lib/server/getAvailability.ts:29]: One less iteration

File: packages/lib/server/getAvailability.ts [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+            ),
+          });
+        }
+      }
+      return dates;
+    }, {} as Record<string, { start: Dayjs; end: Dayjs }[]>);
+
+  const dateOverrides = availability
+    .filter((availability) => !!availability.date)
+    .reduce((dates, override) => {
```
Comment (zomars) [inline packages/lib/server/getAvailability.ts:55]: Same as above

**Review by zomars — CHANGES_REQUESTED** [review]
Codewise looks pretty good. Let's check the override issue and the type issues as well. Nice work @emrysal 🙏🏽

**Comment by grapeslush** [comment]
Hello, just checking again on this fix. Any updates on next steps?

**Comment by mrkirchner** [comment]
Any reason why this pr is still open it’s been almost 2 months.  A few people have been waiting on to see if it fixes a daylight savings issue.

**Comment by emrysal** [comment]
Hi @grapeslush @mrkirchner Yeah this PRs been open for a while, I'm working on closing it off this week; the last 3 weeks have been extremely hectic with other critical issues and investigations.

**Comment by roae** [comment]
I Found this bug, 

I disabled all days, then added 3 date overrides, March 13, 14 and 15 on a 9am to 5pm (America/Detroit) schedule.
imagen

In the calendar it shows me March 12, 13 and 14 enabled.
imagen

**Comment by zomars** [comment]
Drafting until DST bug fixes are applied.

**Pull request closed without merging**